### ISSUE TITLE
More specific exceptions for exceeded socket timeouts

### ISSUE THREAD

**Issue by noobgramming** [eclipse-vertx/vert.x#4675]
#### Describe the feature

It's common to get a Socket Closed exception when exceeding various timeouts (connection timeout, idle timeout, read timeout, write timeout). In my experience it can be frustrating to figure out where these errors are coming from because the stack trace is pruned for performance reasons. Can we switch these out for more specific exceptions to let users know _why_ the socket was closed? 
Ex. `SocketConnectionTimeoutException("Socket connection timed out after X seconds")` and `SocketIdleTimeoutException("Socket timed out from no read/write activity after X seconds")`

#### Use cases

Help users debug cases where default timeouts aren't long enough for their use-cases. Or where timeouts have been improperly configured

#### Contribution

I may be able to implement this with some direction.

**Comment by vietj** [comment]
I agree we should have better reporting, however I believe the error message should be enough than introducing a specific java class for every king of possible timeout.
